### DISCUSSION TITLE
Bad size of marks file

### DISCUSSION THREAD

**Discussion by athree** [ClickHouse/ClickHouse#45748] [Q&A] [answered]
my clickhouse instance got a problem
Bad size of marks file
/myck/b3f/b3f288e0-520b-4d71-9bf9-c576affa3870/20230126_71773_71773_0/data.mrk3  0, must be: 752: While executing MergeTreeReverse. 

it does not appear on just one partition,  list the partition directory, all file seems 0 size;
-rw-r----- 1 root root 0 Jan 26 03:09 checksums.txt
-rw-r----- 1 root root 0 Jan 26 03:09 columns.txt
-rw-r----- 1 root root 0 Jan 26 03:09 count.txt
-rw-r----- 1 root root 0 Jan 26 03:09 data.bin
-rw-r----- 1 root root 0 Jan 26 03:09 data.mrk3
-rw-r----- 1 root root 0 Jan 26 03:09 default_compression_codec.txt
-rw-r----- 1 root root 0 Jan 26 03:09 minmax_ds.idx
-rw-r----- 1 root root 0 Jan 26 03:09 partition.dat
-rw-r----- 1 root root 0 Jan 26 03:09 primary.idx

but the partition in system.parts record a valid size value
![image](https://user-images.githubusercontent.com/10734542/215311579-6511dcdb-76fe-4313-ba58-67d79c991912.png)

it seems write real data failed, but log right; for the written error, i scan /var/log/messages, there is strange thing at the same time

Jan 25 03:08:39 ... kernel: attempt to access beyond end of device
Jan 25 03:08:39 ... kernel: nvme5n1: rw=0, want=3750000008, limit=3750000001

and i find all clickhouse error partition created time is around the 'attempt to access beyond end of device' log time.

my issue machine kernel version is  3.10.0
others machine without this issue version is 4.19.91

so is there any advise for this issue?

**Comment by athree** [comment]
1. my ck version is 22.9;
2. disk is not full;
3. dmesg log rw=0 means read;

**Comment by athree** [comment]
after i set 
min_rows_to_fsync_after_merge = 1
min_compressed_bytes_to_fsync_after_merge = 1
min_compressed_bytes_to_fsync_after_fetch = 1
fsync_after_insert = 1
fsync_part_directory = 1

0 size bad file disappered, so how does it works?

**Comment by mdonkers** [comment, marked as answer]
What you are seeing is just how Linux works with IO.
Until data is really flushed to disk, the writes are only in PageCache (memory). That's why you're seeing the files but still as empty.
By changing the parameters as above, you're telling ClickHouse to explicitly flush (via `fsync`).
Without those settings, files should appear correct eventually as well (and even `fsync` is not 100% reliable, see https://news.ycombinator.com/item?id=19119991)

Why you might not be seeing similar issues on the other system can be because perhaps different configuration or settings in the kernel changed to flush sooner.

The kernel messages should not be related to ClickHouse, and we also haven't received similar issues around this. From what I'm reading it might be a disk configuration issue or corrupted drive. Could you verify everything is fine (check disks etc) and possibly upgrade to a later kernel to check if the problem persists?
